### PR TITLE
Automate keras layer testing

### PR DESCRIPTION
**Pull request by jvmancuso**
Cleaning keras layers and automating their testing

### CODE REVIEW

File: tf_encrypted/keras/layers/convolutional.py [hunk 1]
```diff
@@ -101,29 +103,39 @@ def __init__(self,
     self.strides = strides
     self.padding = padding.upper()
     self.data_format = data_format
+    if activation is not None:
+      logging.info("Performing an activation before a pooling layer can result "
```
Comment (mortendahl) [inline tf_encrypted/keras/layers/convolutional.py]: we should find a better way, but currently you should create a logger object and pass it the right name (see eg config.py)

File: tf_encrypted/protocol/securenn/securenn.py [hunk 2]
```diff
@@ -609,10 +609,6 @@ def _private_compare(prot,
   # TODO[Morten] no need to check this (should be free)
   assert x_bits.backing_dtype == prot.prime_factory
   assert r.backing_dtype.native_type == prot.tensor_factory.native_type
-  print(r.backing_dtype)
```
Comment (mortendahl) [inline tf_encrypted/protocol/securenn/securenn.py]: 😱

File: tf_encrypted/keras/testing_utils.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
+"""Testing utilities for tfe.keras."""
+import numpy as np
+import tensorflow as tf
+
+import tf_encrypted as tfe
+
+
+def agreement_test(layer_cls, kwargs=None, input_shape=None,
```
Comment (mortendahl) [inline tf_encrypted/keras/testing_utils.py]: rename `layer_cls` to `tfe_layer_cls`?

File: tf_encrypted/keras/layers/activation_test.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+class TestActivation(unittest.TestCase):
+
+  def setUp(self):
+    tf.reset_default_graph()
+
+  def test_activation_relu(self):
+    self._core_activation(activation="relu")
+
+  def _core_activation(self, **layer_kwargs):
```
Comment (mortendahl) [inline tf_encrypted/keras/layers/activation_test.py]: can't you inline this function?

File: tf_encrypted/keras/testing_utils.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    ValueError: if `input_data is None and input_shape is None`.
+  """
+  input_shape, input_data = _sanitize_testing_args(input_shape, input_data)
+
+  tf_layer_cls = getattr(tf.keras.layers, layer_cls.__name__)
+  tfe_kwargs = {**kwargs, **tfe_kwargs}
+
+  with tfe.protocol.SecureNN() as prot:
+    tfe_layer = layer_cls(**tfe_kwargs)
+    x = prot.define_private_variable(input_data)
```
Comment (mortendahl) [inline tf_encrypted/keras/testing_utils.py]: you should be able to call `tfe.define_private_variable` and avoid the explicit protocol reference

File: tf_encrypted/keras/testing_utils.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+#   #                        'dtype=%s but expected to find %s.\nFull kwargs: %s' %
+#   #                        (layer_cls.__name__,
+#   #                         x,
+#   #                         keras.backend.dtype(y),
+#   #                         expected_output_dtype,
+#   #                         kwargs))
+
+
+def _sanitize_testing_args(input_shape, input_data):
+  """Fill input_shape or input_data."""
```
Comment (mortendahl) [inline tf_encrypted/keras/testing_utils.py]: docstring could be a bit clearer

File: tf_encrypted/keras/layers/activation_test.py (hunk 4 above)
Comment (jvmancuso) [reply tf_encrypted/keras/layers/activation_test.py]: yes, but this follows the more general construction of testing demonstrated in dense_test.py and convolutional_test.py, which seems preferable when considered in context

**Pull request merged**